### ISSUE TITLE
[4.0][Admin template] metismenu.js issue?

### ISSUE THREAD

**Issue by infograf768** [joomla/joomla-cms#25953]
### Steps to reproduce the issue
Create a custom admin menu with no presets (we have other issues with presets and I will create another issue for that.)
Create a menu module for that menu in the "menu" position.
Make sure the module is placed AFTER the default Admin Menu module in the Manager.

Create a Heading menu item in that custom admin menu.
Create any other menu item as child of that Heading menu item.

The result is that we have both the default Admin Menu and this custom one in the menu side bar. Below, the custom admin menu starts with Global Config.

<img width="266" alt="Screen Shot 2019-08-21 at 10 31 37" src="https://user-images.githubusercontent.com/869724/63416185-47314980-c3ff-11e9-8662-009bf2670cb0.png">

### Expected result
When clicking on the Heading menu item, it should expand.
Same for a Container menu item

### Actual result
It does not expand and no error is displayed.
The event is not added to the element.

We would get the same issue if the Custom Menu is on top of the side bar but this time for all the Menus of the default admin menu which should expand.

Is it metismenu.js issue or somewhere else?

@Fedik

**Comment by Fedik** [comment]
It actually a mod_menu in general, and admin-menu.js.
a mod_menu produces menus with duplicated ids:
![screen 2019-08-21 15 13 13 682x111](https://user-images.githubusercontent.com/1568198/63431356-54fac500-c427-11e9-81a6-4f8307f926e8.png)

and admin-menu.js cannot handle multiple menu, it was coded only for one #menu

![screen 2019-08-21 15 22 14 396x82](https://user-images.githubusercontent.com/1568198/63431455-87a4bd80-c427-11e9-888f-3296491ab7eb.png)

Possible solution:
Drop use of ID, and use a "class" instead


Duplicated ID not allowed anyway.

**Comment by infograf768** [comment]
@Fedik 
I understand. Not sure I can solve myself.

**Comment by Fedik** [comment]
best would be if it will fix who wrote it :)

**Comment by infograf768** [comment]
Looks like metismenu wants only an id
https://onokumus.com/metismenujs/#install

As we are loading the same module twice, we have the same id which is not incremented.
`echo '<ul id="menu" class="' . $class . '">' . "\n";`

**Comment by Fedik** [comment]
> Looks like metismenu wants only an id

hm, not really, it want an element or string selector https://github.com/onokumus/metismenujs/blob/5648f65/src/index.ts#L23

in theory we can change the code from ` new window.MetisMenu('#menu');`
to smething like:
```javascript
var allMenus = document.querySelectorAll('ul.foo-bar-menu-class');
allMenus.forEach(function(menu){
  new MetisMenu(menu);
});
```

It will not depend from ID, but ID still must be incremented, 
Maybe just `id="menu<? echo $module->Id ?>"`
But I not sure about styling, I suspect it may used in CSS, then also will be need changed CSS to use class instead of id.

**Comment by Fedik** [comment]
a suggested fix seems works, but
unfortunately there more code that depend from "only single menu"
https://github.com/joomla/joomla-cms/blob/01600363ada49352df63049587df010b1aaeddb7/build/media_source/mod_menu/js/admin-menu.es6.js#L132-L207

so it not that easy

**Comment by infograf768** [comment]
Modified default.php to use
`echo '<ul id="menu' . $module->id . '" class="' . $class . '">' . "\n";`
then we indeed get a different id per module which anyway solved that aspect.

I then simplified the class code in admin-menu.js to see what happens
```
  var allMenus = document.querySelectorAll('ul.main-nav');

  allMenus.forEach(function(menu) {
  new MetisMenu(menu);
  });
```
tested and it works (after making sure a thousand times that we have no cache anywhere) and also got an error 
`ReferenceError: mainNav is not defined`
because we have further down
`var mainNav = document.getElementById('menu');`
which obviously is wrong as module id is added to `menu`

Therefore it looks like we do have to get the id anyway.

**Comment by Fedik** [comment]
> `var mainNav = document.getElementById('menu');`

this thing also has to be changed to use class, and loop over all menu, or use only first menu.
I just not very understood, for what all that extra code in "Sidebar Nav"

**Comment by infograf768** [comment]
> I just not very understood, for what all that extra code in "Sidebar Nav"

Agree. It looks like working eventhough we have this js error.

**Comment by infograf768** [comment]
@Fedik
I tried another way to get the id

```
    var forEach = function (array, callback, scope) {
      for (var i = 0; i < array.length; i++) {
        callback.call(scope, i, array[i]); // passes back stuff we need
      }
    };

    var allMenus = document.querySelectorAll('ul.main-nav');

    forEach(allMenus, function (index) {
      var x = document.getElementsByTagName('ul')[index];
      var mainNav = document.getElementById(x.id);
      console.log(x.id);
    });
```
But x.id gives both the `menu#` and the ul id of the children ul i.e. `collapse#`

![Screen Shot 2019-08-23 at 11 42 07](https://user-images.githubusercontent.com/869724/63583333-18e87100-c59b-11e9-9bf6-3051a467f159.png)

![Screen Shot 2019-08-23 at 11 40 48](https://user-images.githubusercontent.com/869724/63583236-e8a0d280-c59a-11e9-9fde-a5f4f10065b5.png)

**Comment by Fedik** [comment]
maybe try do [as you did before](https://github.com/joomla/joomla-cms/issues/25953#issuecomment-523899852), and `var mainNav = document.getElementById('menu');` replace to
`var mainNav = document.querySelector('ul.main-nav');`
this will pick a first found menu,
it a bit dirty but also should work, in theory

**Comment by infograf768** [comment]
+1 Indeed, I do not get any more errors.
I will make a PR

**Comment by ghost** [comment]
Closed as having Pull Request #25995
